pages/en/lb4/Decorator.md

```diff
@@ -0,0 +1,125 @@
+---
+lang: en
+title: 'Decorators'
+keywords: LoopBack 4.0, LoopBack-Next
+tags:
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Decorators.html
+summary:
+---
+
+A decorator allows you annotate or modify your class declarations and members with metadata.
+
+## Introduction
+
+*To get familiar with this TypeScript feature, try read [Understand TypeScript Decorator](#understand-typescript-decorator)*
+
+Decorator is a good strategy to keep your code loosely coupled and modifies metadata at runtime. Currently LoopBack core provides built-in decorator functions you can use directly by `@myDecorator`, e.g. Route decorators that introduces in section Route-Decorators, and implements a dependency injection system that can apply a decorator by `@inject('myDecorator')`. 
+
+Some components also exports decorators for you:
+- [authentication](#authentication-decorators)
+- [repository](#repository-decorators)
+
+*Janny: Most examples are documented in a certain concept that a decorator works closely with, I am not sure should I turn this page to something like a reference:*
+
+- some usage are well documented, I linked to certain section
+- some usage can be found in unit test, while integration example is missing, should I create on this page?
+- some have integration test, modify it and add on this page?
+- Do I need to document what metadata each decorator adds to the reflector and document in the reference?
+- I am confused with some decorator can be used by either @mydecorator or @inject(‘mydecorator’), are they equal? 
+
+## Route Decorators
+- @get, @post, @put, @patch, @del
+
+  Register route for a controller by rest decorator, for their usage, refer to [Routing to Controllers](controller.htm#Routing-to-Controllers)
+
+- @api
+
+  Register route for a controller by api decorator, more details and example refer to [Specifying Controller APIs](controllers.htm#Specifying Controller APIs)
+
+- @operation
+
+  Expose a Controller method as a REST API operation. Usage refer to [code](https://github.com/strongloop/loopback-next/blob/debe221ee7eb07b6d6962416b79592c8f13d1e91/packages/core/src/router/metadata.ts#L230-L234)
+
+  *Janny: Need example in doc?*
+
+- @param
+
+  Describe an input parameter of a Controller method. An example can be found in [Writing Controller methods](controller.htm#Writing-Controller-methods), more usage refer to [tsDoc](https://github.com/strongloop/loopback-next/blob/debe221ee7eb07b6d6962416b79592c8f13d1e91/packages/core/src/router/metadata.ts#L252-L272)
+
+  *Janny: Change link to tsDoc page when it's done*
+  
+## Dependency Injection
+
+@inject is a decorator to annotate method arguments for automatic injection by LoopBack IoC container. It can only be applied on a **method parameter** or a **property**. Basiclly it defines a reflector metadata for the target class.
```

Comment:
Several items:
- Don't use the word "basically"
- "It can only be applied on a method parameter or a property" doesn't explain that it doesn't mean you can't _inject_ objects/classes, or that those properties can't be complex objects themselves.
- "It defines a reflector metadata for the target class" tells users that they have to target a class (which is not true) and is too vague. "reflector metadata" doesn't tell us anything about how it works.

Reply:
> "It can only be applied on a method parameter or a property" doesn't explain that it doesn't mean you can't inject objects/classes, or that those properties can't be complex objects themselves.

Sorry I am not sure what do you mean here, according to our code you **cannot** inject a class or a method, see: https://github.com/strongloop/loopback-next/blob/master/packages/context/src/inject.ts#L99

> tells users that they have to target a class (which is not true)

From my understanding the `target` has to be a class or to be more accurate,  an instance of class, you can check the [function]( https://github.com/strongloop/loopback-next/blob/master/packages/context/src/inject.ts#L99), and could you give me an example of `@inject` of which the `target` is not an instance?

Reply:
What I'm saying is that the current phrasing could be misread as saying that you can't resolve class instances with `@inject`, which you can.
The `@inject` decorator itself must be applied to references, as you say, but this isn't what you're saying in the above text.

Reply:
Also, `method arguments` implies a parameter passed into a function/constructor, rather than direct assignment of the target property to the injected value/result.

Reply:
@kjdelisle sorry I am still confused. This is the error msg from code:
> @inject can only be used on properties or method parameters.

If "It can only be applied on a method parameter or a property" is not saying the same thing, I really don't understand but I can change it to exactly "@inject can only be used on properties or method parameters.", is that ok for you?